src/JuMP.jl

```diff
@@ -335,21 +335,123 @@ Base.broadcastable(model::Model) = Ref(model)
-[`direct_model`](@ref).
+ * If JuMP is in `DIRECT` mode (i.e., the model was created using
+   [`direct_model`](@ref)), the backend will be the optimizer passed to
+   [`direct_model`](@ref).
+ * If JuMP is in `MANUAL` or `AUTOMATIC` mode, the backend is a
+   `MOI.Utilities.CachingOptimizer`.
 
-If JuMP is in `MANUAL` or `AUTOMATIC` mode, the backend is a
-`MOI.Utilities.CachingOptimizer`.
+**This function should only be used by advanced users looking to access
+low-level MathOptInterface or solver-specific functionality.**
 
-This function should only be used by advanced users looking to access low-level
-MathOptInterface or solver-specific functionality.
+## Notes
+
+If JuMP is not in `DIRECT` mode, the type returned by `backend` may change
+between any JuMP releases. Therefore, only use the public API exposed by
+MathOptInterface, and do not access internal fields such as
+`backend(model).optimizer.model`. If you require access to the
+innermost optimizer, see [`unsafe_backend`](@ref). Alternatively, use
+[`direct_model`](@ref) to create a JuMP model in `DIRECT` mode.
+
+See also: [`unsafe_backend`](@ref).
 """
 backend(model::Model) = model.moi_backend
 
+"""
+    unsafe_backend(model::Model; attach_caching_optimizers::Bool = true)
+
+Return the innermost optimizer associated with the JuMP model `model`, skipping
+`CachingOptimizer` and `LazyBridgeOptimizer` layers as needed.
+
+If `attach_caching_optimizers`, `MOIU.attach_optimizer(model)` will be called on
+all `MOIU.CachingOptimizer`s in the stack.
+
+**This function should only be used by advanced users looking to access
+low-level solver-specific functionality.**
+
+See also: [`backend`](@ref).
+
+## Unsafe behavior
+
+This function is unsafe because the innermost optimizer may not represent the
+same problem as `model`.
+
+In particular, if you modify the JuMP model, the reference you have to the
+backend may be invalid. Using an invalid reference may result in incorrect
+solutions, errors, or Julia crashing. Always get a new reference by calling
+`unsafe_backend` before calling any low-level functionality.
+
+Moreover, if you pass `attach_caching_optimizers = false`, you should assume
+that any `CachingOptimizer`s may be in the `EMPTY_OPTIMIZER` state.
+
+## Alternatives
+
+Instead of this function, you should consider creating a model using
+[`direct_model`](@ref) and calling [`backend`](@ref) instead.
+
+## Examples
+
+```julia
+model = Model(GLPK.Optimizer)
+@variable(model, x >= 0)
+glpk = unsafe_backend(model)
+# ... use GLPK ...
+@objective(model, Min, 2x + 1)
+glpk = unsafe_backend(model)
+# ... use GLPK again...
+```
+
+Suggested alternative:
+```julia
+model = direct_model(GLPK.Optimizer())
+@variable(model, x >= 0)
+glpk = backend(model)
+# ... use GLPK ...
+@objective(model, Min, 2x + 1)
+# ... use GLPK again. No need to call `backend` twice...
+```
+
+"""
+function unsafe_backend(model::Model; attach_caching_optimizers::Bool = true)
```

Comment:
I would default to `false` for `attach_caching_optimizers`

Reply:
I think we want `true`. Surely if you want to access the low-level model, you want it attached? When would you want it explicitly not attached?

Reply:
I understand that you want it attached when you want to print the model but in these cases, we now have writers in JuMP. In which other cases do you want it not attached ?
You don't need to have it attached when your just setting parameters, ...
If it's opt-in, it will force the user to read the docstring before using this option and understand that it can be automatically and silently detached. I would even prefer not having this option and have the user call `attach_optimizer` before calling `unsafe_backend`.
Having `true` by default can have 2 use cases: 1) saves some keystrokes and 2) allow it to just work for a user not understanding these `CachingOptimizer` layers.
I don't think 1) is so important and for 2), the same users will probably be even more confused when the model is detached and his changes are lost.

Reply:
I guess we could just not have a default...

Reply:
I meant "In which other cases do you want it attached ?"

> I guess we could just not have a default...

That would make sense, we have to force the user to understand this before using `unsafe_backend`

Reply:
Calling low level functionality like presolves, fix-and-solves, conflict, etc. The only time you don't want it attached is setting parameters, which can be done via `set_optimizer_attribute`. People should only be accessing this if they want to do advanced stuff. Setting parameters doesn't really count.

Reply:
I can imagine that most use cases need this but I'm worried that the changes will be dropped silently in `AUTOMATIC` mode. Maybe we could add a warning if the mode is not `MANUAL` ? I don't thing people realize that even setting the starting values of a variable can detach an optimizer. I still think requiring the user to write
```julia
MOIU.attach_optimizer(model)
JuMP.unsafe_backend(model)
```
is clean and prevents any confusion.

Reply:
Also, with JuMP extensions that set optimize hook, attaching is not enough so attaching automatically wouldn't work in all use cases.

Reply:
> the changes will be dropped silently in AUTOMATIC mode

I'm confused as to what the problem is with this. Can you give a code example? The attach, modify the unsafe backend, and then modify the JuMP model and lose their changes to the unsafe backend? This is explicitly warned about in the docstring. Having a manual attach doesn't change this. (They make changes when it isn't attached? calling `optimize!` is going to clear them anyway.)

> Also, with JuMP extensions that set optimize hook

JuMP extensions shouldn't be using `unsafe_backend`. Can you give a code example of a use-case?

The main use case is for people wanting to call the solver functions like write or presolve on a model. In that situation, people would have to write `attach_optimizer`. The other use-case is setting some parameter, although they should use `set_optimizer_attribute`, so this isn't a good example. Until we actually have a use-case where this is a bad thing, I think we should keep it...

Reply:
> The attach, modify the unsafe backend, and then modify the JuMP model and lose their changes to the unsafe backend?

Yes.

> Having a manual attach doesn't change this.

No but since it's the user that attached explicitly then he's responsible for making sure it's not detached.

> JuMP extensions shouldn't be using `unsafe_backend`.

No but the user might use `unsafe_backend` when using an extension, e.g.,
```julia
using SetProg
model = Model()
# ...
@constraint(model, ellipsoid ⊆ square)
unsafe_backend(model)
```
Since the inclusion constraints is transformed in an optimizer hook, it won't be done in `attach_optimizer` hence the backend won't have the constraints associated to this inclusion constraints.
Once the user does
```julia
optimize!(model)
```
then the constraints associated to the inclusion constraints are added which could detach the optimizer.

So by attaching automatically we wouldn't do something that we would recommend to the user. What we should recommend is to attach before calling `unsafe_backend` **and** use `MANUAL` mode so that an error is thrown if the backend is detached when they don't expect it to be. When you are using `unsafe_backend`, you need to control the attachment state of the backend and make sure it is not detached between `unsafe_backend` and `optimize!` so it does not make much sense to be in `AUTOMATIC` mode. Maybe we could even thrown a warning if it's on `AUTOMATIC` mode but that's maybe too much.

Reply:
I've reinforced the warning. 

The solution for things like adding constraints is for users to use `direct_model` instead. 

Users shouldn't be calling `unsafe_backend` with a JuMP extension that sets an optimizer hook prior to `optimize!`.

This use-case covers 99% of what people are going to want to do (get access to the solver. Call a summary C API). For anything else, they should use `direct_model`.

Reply:
I vote for no `attach_caching_optimizers` flag and instead forcing users to write:
```julia
MOIU.attach_optimizer(model)
JuMP.unsafe_backend(model)
```
First of all, `unsafe_backend` looks like a getter that doesn't change the state of the JuMP model. It's a bit surprising that it does by default. Second of all, suppose I'm reading this code that someone else wrote, and I haven't read the long docstring for `unsafe_backend`. The call to `MOIU.attach_optimizer` tells me that something state-changing is happening.

I don't really see the value in saving users one line of code. This is already an advanced and complex operation. And as @blegat mentioned, we can't even cover cases like optimizer hooks.

Reply:
And we can the two lines:
```julia
MOIU.attach_optimizer(model)
JuMP.unsafe_backend(model)
```
In the docstring.